### PR TITLE
Brettski/95/new asset type

### LINKED ISSUE #93: Changed speakers return type to PublicProfile from Profile
**Schema Change**

me > profile > MySession > Speakers
- is [Profile]
- should be [PublicProfile]

**note**: this doesn't change data paths only the fields available under speakers

### PR DESCRIPTION
**Pull request by brettski**
wip
fixes #93
add Asset type and resolvers
package bumps
refactor file locations for better organization

### CODE REVIEW

**Comment by csell5** [comment]
sadly the typedefs/datatypes/admin types feel like they should have an interface on them just to keep them all aligned.

**Comment by brettski** [comment]
@csell5 does it matter that they were not part of this change?

**Comment by csell5** [comment]
> @csell5 does it matter that they were not part of this change? 
 

na, I just see that being out of sync by one field.

File: src/graphql/typeDefs/mutations/asset.graphql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+type AssetMutation {
+  "Updates referenced asset"
+  update(asset: AssetUpdateInput!): Asset
```
Comment (csell5) [inline src/graphql/typeDefs/mutations/asset.graphql:3]: this should have a security role.

File: src/graphql/typeDefs/mutations/assets.graphql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+type AssetsMutation {
+  "Creates a new Asset"
+  create(asset: AssetCreateInput!): Asset
+
+  "Asset specific mutations"
+  asset(id: ID!): AssetMutation
```
Comment (csell5) [inline src/graphql/typeDefs/mutations/assets.graphql:6]: this should have a security role

File: src/graphql/typeDefs/mutations/assets.graphql [hunk 3]
```diff
@@ -0,0 +1,10 @@
+type AssetsMutation {
+  "Creates a new Asset"
+  create(asset: AssetCreateInput!): Asset
```
Comment (csell5) [inline src/graphql/typeDefs/mutations/assets.graphql:3]: this should have a security role

File: src/graphql/typeDefs/mutations/asset.graphql (hunk 1 above)
Comment (brettski) [reply src/graphql/typeDefs/mutations/asset.graphql:3]: messy, messy, messy. On it

File: src/graphql/typeDefs/dataTypes/asset.graphql [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  startTime: Date
+  stopTime: Date
+
+  "Entities this asset is assigned to"
+  assignments: [Entity]
+
+  "Owner of asset, entity which manages asset (Partner, Community, Member, Event, etc.)"
+  owner: Entity
+
+  # -- edit information
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/asset.graphql:28]: shouldn't this be required?

File: src/graphql/typeDefs/dataTypes/asset.graphql [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  createdAt: Date
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/asset.graphql:29]: shouldn't this be required?

File: src/graphql/typeDefs/dataTypes/asset.graphql [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  createdBy: PublicProfile
+  lastUpdatedAt: Date
+  lastUpdatedBy: PublicProfile
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/asset.graphql:32]: shouldn't this be required?

File: src/graphql/typeDefs/dataTypes/asset.graphql [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  "Entities this asset is assigned to"
+  assignments: [Entity]
+
+  "Owner of asset, entity which manages asset (Partner, Community, Member, Event, etc.)"
+  owner: Entity
+
+  # -- edit information
+  createdAt: Date
+  createdBy: PublicProfile
+  lastUpdatedAt: Date
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/asset.graphql:31]: shouldn't this be required?

File: src/graphql/typeDefs/dataTypes/asset.graphql [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+  "Entities this asset is assigned to"
+  assignments: [Entity]
+
+  "Owner of asset, entity which manages asset (Partner, Community, Member, Event, etc.)"
+  owner: Entity
+
+  # -- edit information
+  createdAt: Date
+  createdBy: PublicProfile
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/asset.graphql:30]: shouldn't this be required?

File: src/graphql/typeDefs/dataTypes/asset.graphql [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  priority: String
+  startTime: Date
+  stopTime: Date
+
+  "Entities this asset is assigned to"
+  assignments: [Entity]
+
+  "Owner of asset, entity which manages asset (Partner, Community, Member, Event, etc.)"
+  owner: Entity
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/asset.graphql:26]: shouldn't this be required?

File: src/graphql/typeDefs/dataTypes/session/mySession.graphql [hunk 10]
```diff
@@ -3,7 +3,7 @@ type MySession @key(fields: "id") {
   id: ID!
   eventId: ID!
   slug: Slug!
-  speakers: [Profile]
+  speakers: [PublicProfile]
```
Comment (csell5) [inline src/graphql/typeDefs/dataTypes/session/mySession.graphql:6]: shouldn't this be required?

File: src/graphql/typeDefs/mutations/asset.graphql (hunk 1 above)
Comment (brettski) [reply src/graphql/typeDefs/mutations/asset.graphql:3]: Ah, now I remember, not sure what to set it to. Member perhaps

Comment (brettski) [reply src/graphql/typeDefs/mutations/asset.graphql:3]: Nope, never mind. It has to be admin until some of the other security stuff is added

File: src/dataSources/cloudFirestore/asset.js [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  async function setAssignmentsBatch({
+    firestoreBatch,
+    user,
+    assignments,
+    assetId,
+  }) {
+    dlog('set assignments batch %o', assignments);
+    // TODO: verify that user may assign asset to given entity
```
Comment (csell5) [inline src/dataSources/cloudFirestore/asset.js:77]: we should log a github task for this and below so we don't loose track of it.

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 4 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/asset.graphql:28]: No, not required as the default is the current user if not provided. I will add to documentation

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 5 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/asset.graphql:29]: No, not required as the default is the current user if not provided. I will add to documentation

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 4 above)
Comment (csell5) [reply src/graphql/typeDefs/dataTypes/asset.graphql:28]: but this is the asset, so won't we have an owner regardless?

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 5 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/asset.graphql:29]: Ah this is the asset type, so always returned. Yes, okay, required

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 6 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/asset.graphql:32]: Not sure, which one, createdAt or all four? We want to guarantee a value for these?

File: src/graphql/typeDefs/dataTypes/session/mySession.graphql (hunk 10 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/session/mySession.graphql:6]: I guess, that is old code there

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 4 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/asset.graphql:28]: What I said on the other comment for this

File: src/graphql/typeDefs/dataTypes/session/mySession.graphql (hunk 10 above)
Comment (brettski) [reply src/graphql/typeDefs/dataTypes/session/mySession.graphql:6]: @csell, this makes sense, but it should then be added to all the session types as well, only mySession

File: src/dataSources/cloudFirestore/asset.js (hunk 11 above)
Comment (brettski) [reply src/dataSources/cloudFirestore/asset.js:77]: done. #98

File: src/graphql/typeDefs/dataTypes/asset.graphql (hunk 6 above)
Comment (csell5) [reply src/graphql/typeDefs/dataTypes/asset.graphql:32]: I was thinking all 4

File: src/graphql/typeDefs/dataTypes/session/mySession.graphql (hunk 10 above)
Comment (csell5) [reply src/graphql/typeDefs/dataTypes/session/mySession.graphql:6]: yea prolly a good change to make everywhere when we run across it.

**Review by csell5 — CHANGES_REQUESTED** [review]

**Comment by brettski** [comment]
Not sure what happened to the conversation thread on this....

> sadly the typedefs/datatypes/admin types feel like they should have an interface on them just to keep them all aligned.

typedefs/datatypes/admin directory contains inputs only, not types, so they wouldn't be inherited by the interface. Which field did you see missing?

**Review by csell5 — APPROVED** [review]

**Pull request merged**